### PR TITLE
ashift rework

### PR DESCRIPTION
**Pull request by AlicVB**
This is the last part of the c&r "rewrite" #8834

This handle rotation and keystone directly in perspective module : 
- UI rewrite to advertise rotation, as it will certainly be the most used feature...
- rename the module as "rotation and perspective"
- port the c&r feature "straightening drawing line"
- change the rotation slider max bound (available by typing numbers) to -180/180
- allow to choose between 3 perspective modes : none/auto/drawn
- "hide" detailed values under an expander, like what is done in crop module

**currently the "drawing perspective" mode is not done at all** : just the icons are here ! @aurelienpierre : if you have time to look into it, it would be great :) (I have no idea how to translate c&r implementation, which is done in a "big" weird matrix into vertical/horizontal lenshift and shear values, as it's done for automatic mode here)

some screenshots : 
default UI : 
Capture d’écran_2021-07-08_11-54-41
with perspective : 
Capture d’écran_2021-07-08_11-55-14
Capture d’écran_2021-07-08_11-58-34
Capture d’écran_2021-07-08_11-58-55

### CODE REVIEW

**Comment by elstoc** [comment]
Just some minor suggestions about naming for now: I think "drawn" would be better than "drawing" for the method, and perhaps "perspective parameters" (or even just "parameters"?) instead of "perspective values".

**Comment by AlicVB** [comment]
good news, finally, I've reused the actual ashift code completely (and without any change) to implement the "drawing" perspective.
This has the big advantage that the code and the gui are way simpler.
Capture d’écran_2021-09-07_22-15-54


drawn perspective has some little changes comparing to the actual c&r implementation. Mainly to mimic the actual ashift perspective worflow and to keep the code as simple as possible : 
- to apply the perspective, we need to use one of the 3 "fit" icons. No more big "ok" button in the overlay
- overlay are drawn the same way as for the auto mode (colors, ...) only the round at the corner have been added

for instance, I've not implement the "line" move (only the corner move) neither the "keep symmetry" feature (the infinite icons on the overlay in c&r)
If there's real need for this, we can always add it after (but I'm not sure as I found that they are half-broken in actual c&r and nobody raise any issue)

I think I've port all the needed feature now, so I remove the draft mode... But I'm sure this will still need some polishing before merging ;)

**Comment by elstoc** [comment]
Looks good. Given that the automatic model can work from multiple auto-detected horizontal and vertical lines, I was wondering: would it be better to just allow the user to draw their own lines (perhaps lots of them) rather than a single fixed "rectangle" and then let the auto algorithm work the same for both methods.

**Comment by AlicVB** [comment]
@elstoc : Why not... but I would better see it as a third mode.

This will not be so simple to implement, especially because we might have not converging lines, intersection point outside image area, etc...

**Comment by elstoc** [comment]
Could you have the algorithm do its automatic discarding of inconsistent lines (or whatever it uses to ignore things)? I don't really know how it works but I do like the idea of just drawing lines - seems faster/simpler, at least from a usage perspective.

**Comment by AlicVB** [comment]
hmmm... you may be right... I admit that I've not understood all the ashift internals, but effectively the algorithm discards some unrelevant or problematic lines at some point, so we should be able to reuse that...

and I agree that this would be handy to directly draw lines...

That said, if you agree that the current "rectangle" implementation should be kept, I would prefer to see that in another FR/PR as this one is big enough imho...

**Comment by elstoc** [comment]
IMO it would be better to just have one method. The rectangle way is fine but I think direct line drawing would be better if possible. Just my opinion though and I'm not bothered if you want to have two options. I quite often use the existing keystone correction on images where there are only a couple of usable horizontal reference lines and it can take a while to move corners of a pre-drawn rectangle compared to just drawing the lines myself.

BTW I've just noticed that the right-click+drag for rotation interferes with a pre-existing implementation of that key/mouse combination. It's currently used to tell the algorithm to ignore some detected horizontal/vertical structure lines. It's probably best to disable that shortcut whenever there are any on-screen structure lines displayed.

File: src/iop/ashift.c [hunk 1]
```diff
@@ -4772,16 +5184,19 @@ void gui_init(struct dt_iop_module_t *self)
                                              "ctrl+click to only fit rotation\n"
                                              "shift+click to only fit lens shift\n"
                                              "ctrl+shift+click to only fit rotation and lens shift"));
-  gtk_widget_set_tooltip_text(g->structure, _("analyse line structure in image\n"
+  gtk_widget_set_tooltip_text(g->structure, _("automatically analyse line structure in image\n"
                                               "ctrl+click for an additional edge enhancement\n"
                                               "shift+click for an additional detail enhancement\n"
                                               "ctrl+shift+click for a combination of both methods"));
+  gtk_widget_set_tooltip_text(g->draw_structure, _("define manually perspective rectangle"));
```
Comment (elstoc) [inline src/iop/ashift.c:5191]: Can you change this to "manually define perspective rectangle" please

**Comment by AlicVB** [comment]
> Can you change this to "manually define perspective rectangle" please

done. hopefully you are here to fix my english ;)


> BTW I've just noticed that the right-click+drag for rotation interferes with a pre-existing implementation of that key/mouse combination.

good catch. fixed too




> Oh and one other thing.. do we really want to hide the correction buttons underneath a collapsing section? Surely they're key parts that we want to always show.

This has been done on purpose : as this module will replace completely c&r, the rotation becomes the key feature of the module. Most people will only use that, especially (but no only) newcommers. This allow to keep the ui small and simple for them...
Anyway, the expanded state of the perspective section should be remember, so it's not really a big deal either for those who prefer the full interface.

I've thought to your "direct draw" proposal too. I'll give it a try asap...

Thanks for your review and test !

**Comment by AlicVB** [comment]
@elstoc : last commit implement the "direct drawing" you suggest. Finally it was quite simple...

For now, I've done it as a third method. I'm still not sure we can drop the "rectangle" one... as the "direct drawing" as some limitations too, especially for discoverability...

There's still some work to polish all this, esp. I'd like to allow points move for direct drawn lines. But it would be great if other devs can test this and give their opinions...
Thanks !

**Comment by elstoc** [comment]
Works well for me thanks. Just a minor thing... The five "structure" icons look a bit misaligned below the three "fit" icons. I understand that they're separate logical groups of 3 and 2 but it doesn't look quite right. Perhaps either put the X and the eye on a separate line or spread the five icons evenly. If the latter, the label should probably just say "structure".

**Comment by AlicVB** [comment]
- you can now refine manually drawn line
- ui disposition has been a little changed in order to : 
1. get a more balanced buttons postions
2. reflect the order of the actions : first get the structure, then apply the changes

Capture d’écran_2021-09-08_22-24-01

**Comment by elstoc** [comment]
A few more things...

- Much prefer the icon layout now. It never really made sense to have the fit icons first and the new layout is much more logical.
- It's great to be able to adjust the manual lines, but the auto-generated lines also now have adjustment end nodes and, in complex images, you can't really see the lines any more.
- Right-click marks lines as "to be ignored" in auto mode but in manual mode, imo, they should be deleted entirely when right-clicked (like a drawn shape).
- Similarly a couple of times I've tried to click and drag on the line itself to move it but that creates a new line. Again, for people used to drawn shapes it's a habit that will be hard to break.

**Comment by AlicVB** [comment]
Thanks again for your tests.
Last commits implement what you have raised/proposed : 
- fix bug with corner drawing
- use toggle buttons for structure icons
- hide unselected lines in drawn mode
- allow to move drawn line entirely

**Comment by elstoc** [comment]
This is looking really nice. Still a couple of minor comments... 

- In darktable 3.6, clicking repeatedly on the analyse button re-analyses the image and often causes different lines to appear. In this iteration clicking the button a second time makes the lines disappear and reappear (without re-analysing?). I'm not sure which behaviour is best but doesn't appear to be possible to re-analyse the image without exiting and re-entering the darkroom. Perhaps this button should retain the previous behaviour.
- If you click on the analyse button, reset the module parameters and then click analyse again, a message "data pending - please repeat" appears. You have to collapse and re-expand the module to fix this.
- Clicking repeatedly on the new lines and rectangles buttons seems to toggle the lines/rectangle off and forget their shapes / positions. At least while the image remains in the darkroom these lines should be retained. Ideally it would be great if we could keep them between sessions (like C&R does).
- Collapsing the module should un-toggle the edit buttons and take the module out of any current edit mode. Similarly for the "remove line structure information" button and probably the "show/hide" button too

**Comment by AlicVB** [comment]
>     * In darktable 3.6, clicking repeatedly on the analyse button re-analyses the image and often causes different lines to appear. In this iteration clicking the button a second time makes the lines disappear and reappear (without re-analysing?). I'm not sure which behaviour is best but doesn't appear to be possible to re-analyse the image without exiting and re-entering the darkroom. Perhaps this button should retain the previous behaviour.

desactivating/reactivating the button reanalyze the image. No change here... Yes, now the "multiple analyze" need more click... on the other hand, this make all the structure buttons works the same way and I'm not sure that this is a really common case. What need to be handle, thought, is the case of "ctrl/shift click" after a simple click...



>     * If you click on the analyse button, reset the module parameters and then click analyse again, a message "data pending - please repeat" appears. You have to collapse and re-expand the module to fix this.

this is already the case in the current implementation. This need to be fixed, but I would prefer to do it in separate PR, to keep things cleaner...



>     * Clicking repeatedly on the new lines and rectangles buttons seems to toggle the lines/rectangle off and forget their shapes / positions. At least while the image remains in the darkroom these lines should be retained. Ideally it would be great if we could keep them between sessions (like C&R does).

that would mean to "store" the lines positions in module parameters. Why not... I need to check, but iirc, that would have the drawback that we need to have a fixed max number of lines... But that shouldn't be a real problem imho



>     * Collapsing the module should un-toggle the edit buttons and take the module out of any current edit mode. Similarly for the "remove line structure information" button and probably the "show/hide" button too

good point.

Thanks again for your tests and review. I'll try to have a look at that asap... Not sure to have lot of time this week, but...

**Comment by AlicVB** [comment]
I think I've fixed all the points mentioned by @elstoc  except the last one. After some tests, I'm not sure it's a problem at all. I even fine the behavior quite handy as if you switch to another module for whatever reason, you find the module in it's exact previous state when you come back (or I may have missed something)

Note : due to the saving of drawn structure in parameters, the module version has been updated, so before testing don't forget to keep a backup of your edits first ;)

**Comment by elstoc** [comment]
With the toggle buttons, do we still need the show/hide lines button? 

If I draw a rectangle, then draw some lines, then click the auto-detect button and finally toggle all of the drawing modes off, which one should it use to do the actual correction? The module still seems to recall the drawn lines and the drawn rectangle but seems to only use the auto-detection to do the correction in this scenario.

Should we make it so that whichever button you click disables the others and deletes any unrelated lines? Perhaps we can highlight the mode that's in use, maybe make the icons bold or something? I think maybe in all my suggestions I haven't taken this nuance into account.

**Comment by AlicVB** [comment]
> With the toggle buttons, do we still need the show/hide lines button?

This has the only advantage that it reset the eventually saved drawn lines. It will need to update the tooltip !



> An alternative might be to disable the fit buttons if none of the structure buttons are toggled.

I tend to prefer that solution... Will do it too !

**Comment by elstoc** [comment]
> As usual, thanks for your tests. I'm really impressed of your reactivity and your ability to find issues and UI problems !

My pleasure. I can hit buttons randomly until stuff breaks with the best of them! My reactivity is because this is a really great addition and I will be very happy to see the back of crop & rotate. Great work on this :)

**Comment by AlicVB** [comment]
- fit buttons are now disabled if no structure button is toggled.
- for the clean button, I've finally remove it entirely as in fact it was not working as expected and is not really useful (as you said)

**Comment by elstoc** [comment]
> for the clean button, I've finally remove it entirely as in fact it was not working as expected and is not really useful (as you said)

I was actually referring to the other button. The "toggle visibility of structure lines" seems unnecessary because you can just toggle off the appropriate structure button. We can probably do away with both - users can reset the module to clear any drawn lines, and use the toggle buttons to hide structure lines.

Apart from that I think I've run out of ways to break this now. @TurboGit IMO this is ready for reveiw/merge.

**Comment by AlicVB** [comment]
@elstoc : tbh, even in actual version, I've never understand the utility of the "visibility" toggle. If we all agree to remove it I'll do it. It will remove some code complexity (even if for this button, it's not a really big issue)

**Comment by elstoc** [comment]
Let's go for it. It's just confusing now.

**Comment by AlicVB** [comment]
- visibility icons removed :)
- when I was at it, I've also removed the ability to move (resize ??) the crop area in "original image" crop mode : in real use, I find it completely unpredictable : the size is changing as well as the position, it's almost impossible to come back to the original state, ...
If one wants to adjust more precisely the crop area, better use the crop module ! the only advantage to offer crop here is that we can do it automatically as we know the maths used to change the image edges !

**Comment by elstoc** [comment]
I think you can probably remove "[WIP]" from the description now

**Comment by amiryal** [comment]
The old c&r let you right-click-and-drag to automatically rotate, so that the drawn line becomes either fully horizontal or fully vertical. From reading this thread, it seems like there is no direct equivalent in the new implementation. Is it possible to perform an automatic fit with “rotate only” (without any keystone adjustment) in this new module?

**Comment by elstoc** [comment]
The right-click-drag to rotate has been ported to this module

**Comment by TurboGit** [comment]
@AlicVB : The support for moving the crop area and automatically follow the constraints is not working anymore.

**Comment by TurboGit** [comment]
> IMO this should just allow "crop to largest visible rectangle" and allow for creative crop in the crop module

This is not an acceptable regression to me. I do use the resize/crop on perspective correction module and do not need to use another module just for that. Indeed, hard to get the original area in current module, we can add an option for that or don't care as I have never had the need for this.

**Comment by AlicVB** [comment]
@TurboGit : As said, I *find* the implementation buggy so I have preferred to take the opportunity to cleanup the code... but your experience means it's still usable, and as it was here before... If one is annoy by this, he will open an issue ;)

Reenabling the feature is just a matter of dropping the last commit. Do you want me to do it ?

**Comment by TurboGit** [comment]
> Reenabling the feature is just a matter of dropping the last commit. Do you want me to do it ?

Yes please.

BTW, what's was broken in this feature? As I said I have used it quite a lot without issue...

**Comment by AlicVB** [comment]
@TurboGit : done.

The reason why I find the current implementation not useful in real world is that most of the times (in my usecases) what I want is to move slightly the crop area to the left, right, top, ... in order to not hide or cut an important  thing in the image. While keeping the maximum image size. That's almost impossible as each mouse move as a very important effect, especially by resizing the crop area. Another issue is that getting back to the original "optimal" crop is almost impossible without unselect autocrop then select it again (that's due to the fact that you need to move back the mouse at the exact "gravicenter" of the image, which is not marked) Lats, as ashift is before crop in the pipe, you can't readjust the crop by extending it in the crop module...

Now that we want to make ashift the default module for rotations, that's not so great imho.

At first I have thought about enhancing this feature, but when you thought at this, you  duplicate the crop module and are back to the unmaintainable c&r monster :(

That's said, I won't oppose to keep the current feature, as it has a very small impact on code complexity, but I would prefer that we keep in mind that we should refrain of adding more features/complexity here !

**Comment by TurboGit** [comment]
Ok thanks for putting this back. Now I'm wondering if we should not always display the perspective controls. Today it is collapsed, but as we are going to deprecate crop&rotate the perspective correction could well be the most used feature in this module, no? And the module is not that big. So maybe just remove the folding for perspective?

Anyway, this is working fine so I'll merge to get more field testing. We can handle this folding (if needed) in another PR.

Last but not least, thanks a lot @AlicVB  for the hard work on this module !

**Comment by TurboGit** [comment]
In fact no, I still have an issue. The automatic crop is ok. But I cannot use the manual structure or using a rectangle to define perspective now. That is, only the 3rd button in the structure line is working. Using the first two, when hovering the image it seems that we are in the lines adjustment mode (adding or removing auto detected lines).

**Review by TurboGit — CHANGES_REQUESTED** [review]
Only the fully auto mode works.

**Comment by AlicVB** [comment]
> as we are going to deprecate crop&rotate the perspective correction could well be the most used feature in this module, no?

no, I'm quite sure that the most used feature will be the rotation... and perspective adjustments will only be used in few cases (except for architecture pictures, or such special cases, maybe)

Anyway, the fold/unfold status is remember across sessions, so people have the choice to keep it expanded or not...



> In fact no, I still have an issue. The automatic crop is ok. But I cannot use the manual structure or using a rectangle to define perspective now. That is, only the 3rd button in the structure line is working. Using the first two, when hovering the image it seems that we are in the lines adjustment mode (adding or removing auto detected lines).

hmmm... I've never seen that (and I can't reproduce). Any special steps to reproduce the issue ? I'll also recheck the code, as it looks like a missing variable reset...

**Comment by TurboGit** [comment]
Ok, found the issue. This is due to the preference "reduce resolution of preview images" in "darkroom". If set to original all is OK, if set to any other value it is NOK. This means that some computation should be using the down-sampling factor. See` pr_d` constant used in many iop.

**Comment by TurboGit** [comment]
BTW, this explain why I add no issue on the other computer where I tested this the first time as I do not have down-sampling activated.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Fix when using down sampling.

**Comment by AlicVB** [comment]
@TurboGit : thanks a lot for the debugging !
I should have fixed your issue. The most complex was the parameters in case one changed the preview downsampling in the middle... but I think we are ok now (better double check)
Thanks a lot !

**Comment by TurboGit** [comment]
@AlicVB : Ok the issue with preview down sampling s fixed. I have found two other issues though.

1. reset module
2. select the perspective rectangle
3. move bottom-right corner a bit, the right line is now green (as for vertical aligment)
4. click on the horizontal correction button (middle) or the horiztonl+vertical correction buttpn (right) you get a message: `not enough structure for automatic correction`.

If in 3 you don't move the corner the alignment is done.

BTW, for portrait image the correction is done with a rotation if you don't move the corner. That's also something we want to fix.

**Review by TurboGit — CHANGES_REQUESTED** [review]
Two more issues found.

**Comment by AlicVB** [comment]
@TurboGit : the 2 issues seems to comes from the "reset" bug mentioned earlier in this thread : 

For some reason, when you click on the reset button, the preview is not reprocessed... well no call to ashift process routine is done at least. and this (ashift process routine) is needed to get the preview buffer and various other infos, including real orientation...

What is surprising is that if you click on the module header to "fold" the module, then the preview is reprocessed...

I'm quite sure there's some cache invalidation to do in `gui_reset`, but I fail to find how to do that for now...

Maybe you have an idea ? Thanks

**Comment by AlicVB** [comment]
ahah finally, I find it ! your 2 last bugs are fixed, and the "reset" one too !

waiting for the next issue :)

thanks again for your tests and review

**Comment by TurboGit** [comment]
> waiting for the next issue :)

I'm sorry, you'll be very disappointed but I was not able to find another issue :)

**Review by TurboGit — APPROVED** [review]
All good to me now. Thanks again for the hard work @AlicVB !

**Comment by elstoc** [comment]
Last thing is to deprecate Crop & Rotate then!

**Comment by TurboGit** [comment]
> Last thing is to deprecate Crop & Rotate then!

Yes, was waiting for a bit more field testing to ensure nothing is broken, then we'll deprecate.

**Comment by elstoc** [comment]
Talking of which, we have been discussing on another issue the possibility of adding a new tag, perhaps something like "field-testing required" so that testers using master know what features have been merged but need further testing.

**Comment by s7habo** [comment]
> Talking of which, we have been discussing on another issue the possibility of adding a new tag, perhaps something like "field-testing required" so that testers using master know what features have been merged but need further testing.

I'm in the process now and what I can say so far, you guys have done an excellent job. Perspective correction is so much easier now!
Chapeau!

**Comment by s7habo** [comment]
Some small comments by testing the module:

If you _only want to rotate the photo_, the line could be helpful to trace the horizon for example (in this case I moved it down a bit so that it becomes visible): 

Perspective1

The problem is that this is _not enough_ to make the correction. Moreover, it is not immediately obvious _where_ to activate the correction.  Perhaps "fit" label should be replaced with "activate". When I then click on one of the "fit" options, I get "_not enough structure for automatic correction_". 

Perspective2

Only when I have drawn an additional horizontal and two vertical lines, the correction is activated when I click on one of the "fit" options. However, I had to estimate the vertical lines, because I have no reference points for them in the photo:

Perspective3

From previous tests of use, my suggestion would be to be able to activate the correction _already with a single line_ (horizontal or vertical). In this case, it becomes clear that the correction here concerns _only the rotation_. 

What I also noticed is that the rotation function is _separated_ from the other functions of the perspective correction. And in the current design the _automatic cropping is bound to the rotation function_. And by separating the rotation from the perspective correction, it is _not obvious_ that automatic cropping _also refers to the perspective correction_.

In my experience, the rotation function could well become the part of other perspective correction options, and automatic cropping can then be applied as separate option to all functions together.

**Comment by amiryal** [comment]
@s7habo I also had the same intuition, that the hand-drawn structure with a single line would be used to calculate automatic rotation. JFYI, as **elstoc** had already pointed out to me:
> The right-click-drag to rotate has been ported to this module

I am not a UX expert, so maybe someone can come up with a better suggestion, but I also think this one is good:
> my suggestion would be to be able to activate the correction _already with a single line_ (horizontal or vertical). In this case, it becomes clear that the correction here concerns _only the rotation_.

Now, here is a question: does it make sense to do both perspective corrections _and_ a separate right-click-drag rotation? If not, then I would also suggest to cause right-click-drag to draw an editable line of structure (instead of disappearing once the rotation angle is set).

**Comment by elstoc** [comment]
Perhaps better to raise a FR to discuss further changes now.

**Comment by s7habo** [comment]
> @s7habo I also had the same intuition, that the hand-drawn structure with a single line would be used to calculate automatic rotation. JFYI, as **elstoc** had already pointed out to me:
> 
> > The right-click-drag to rotate has been ported to this module

This function is very fast and effective but hidden. And you need to know the keyboard shortcut. 

If the rotation were part of the other functions of the perspective correction, and the same tools (in this case the line) could be used as for other corrections, then the whole functionality would be more consistent.

> Perhaps better to raise a FR to discuss further changes now. IMO, though, it's nicer not to have to draw a line and _then_ click a button to correct rotation.

That is only cosmetic change. The module as it is developed now works perfectly. As far as I am concerned, it can stay like this.

**Comment by AlicVB** [comment]
@s7habo : thanks for your tests and comments. as said by @elstoc it would have been simpler for us to open a new issue/FR, that ensure your comments are not lost...
To answer quickly, 
- right-click + drag is an hidden feature, but that was also the case in c&r. It has proven it's efficiency !
- I would prefer to avoid using "perspective" lines for rotation, as the algorithm is not designed for that.
- as the "lines" icons is under the "perspective helpers" section, I don't find that confusing imho
- rotation slider is a little "apart" from the other slider, because lot of user will only use it and don't touch at the perspective part. But yes, automatic perspective correction will change the rotation too. We can really avoid that !

I've also proposed a new PR to enhance tooltips, esp. for the right-click+drag feature (see #10101)

**Comment by elstoc** [comment]
should we recommend (in the documentation) that the user sets rotation first and then perspective correction?

**Comment by AlicVB** [comment]
> should we recommend (in the documentation) that the user sets rotation first and then does perspective correction?

No, I don't think so, maybe even the contrary, as the fit algorithm will change the rotation in any case. Then you can eventually adjust it to your needs.
And for the auto-structure detection, the algorithm use the buffer at the module entry, so rotation or any other parameters doesn't change anything.

**Pull request merged**